app/processing/siti-agri/handle-known-defects/cap-bps-penalties/recalculate-bps-penalties.js

```diff
@@ -0,0 +1,34 @@
+const recalculateBPSPenalties = (paymentRequest) => {
+  const schemeCodes = [...new Set(paymentRequest.invoiceLines.map(invoiceLine => invoiceLine.schemeCode))]
+
+  schemeCodes.forEach((schemeCode) => {
+    const invoiceLinesByScheme = paymentRequest.invoiceLines.filter(invoiceLine => invoiceLine.schemeCode === schemeCode)
+    let grossAfterPenalties = calculateGrossAfterPenalties(invoiceLinesByScheme)
+
+    if (grossAfterPenalties < 0) {
+      const p04Penalty = invoiceLinesByScheme.filter(invoiceLine => invoiceLine.description.match(/^P04/gm))[0]
+      const p02Penalty = invoiceLinesByScheme.filter(invoiceLine => invoiceLine.description.match(/^P02/gm))[0]
+
+      if (p04Penalty) {
+        p04Penalty.value = p04Penalty.value - grossAfterPenalties > 0 ? 0 : p04Penalty.value -= grossAfterPenalties
+        grossAfterPenalties = calculateGrossAfterPenalties(invoiceLinesByScheme)
+      }
+
+      if (grossAfterPenalties < 0) {
+        p02Penalty.value -= grossAfterPenalties
+      }
```

Comment:
Does this need to be surrounded by this if? It looks like to me that it's saying the same thing as the outer if

Reply:
Yep, It is the same as the outer `if` however i think it is needed.

On line 6 we calculate the `grossAfterPenalties` for the first time
On line 12 we check if there is a `p04Penalty`, if there is we update the penalty value, after this we need to recalculate the `grossAfterPenalties` on line 14.

We need the `if` on line 17 to check if the grossAfterPenalties is still < 0 after the recalculation on line 14, if it is we need to also reduce the `p02Penalty`

So there is some duplication but couldnt think of a better way to do it, possibly some kind of while loop could work. Do you have any suggestions?